iroha/src/kura.rs

```diff
@@ -180,126 +189,249 @@ impl Default for Mode {
+    /// Returns a sorted vector of blocks starting from 1 height to the top block.
     #[iroha_futures::telemetry_future]
-    pub async fn read(&self, height: u64) -> Result<VersionedCommittedBlock> {
-        let path = self.get_block_path(height);
-        let mut file = File::open(&path).await.wrap_err("No file found.")?;
-        let metadata = metadata(&path).await.wrap_err("Unable to read metadata.")?;
-        #[allow(clippy::cast_possible_truncation)]
-        let mut buffer = vec![0; metadata.len() as usize];
-        let _ = file.read(&mut buffer).await.wrap_err("Buffer overflow.")?;
-        VersionedCommittedBlock::decode_versioned(&buffer)
-            .wrap_err("Failed to read block from store.")
+    pub async fn read_all(&self) -> Result<Vec<VersionedCommittedBlock>> {
+        let mut blocks: Vec<VersionedCommittedBlock> = Vec::new();
+        let base_indices = sorted_storage_files_base_indices(&self.path).await?;
+        for base_index in base_indices {
+            let path = self.path.join(base_index.to_string());
+            let mut read_blocks = read_all_from_storage_file(&path)
+                .await
+                .wrap_err("Failed to deserialize blocks.")?;
+            let prev_block = read_blocks
+                .first()
+                .wrap_err(format!(
+                    "No blocks were found in file {}. Data storage might be corrupted.",
+                    path.to_string_lossy()
+                ))?
+                .header()
+                .height;
+            if let Some(last_block) = blocks.last() {
+                let new_block = last_block.header().height;
+                if prev_block != new_block + 1 {
+                    return Err(eyre::eyre!("Incosequential blocks were found: [{} - {}]. Data storage is probably corrupt.", prev_block, new_block));
+                }
+            } else if prev_block != 1 {
+                return Err(eyre::eyre!(
+                    "First block not found. Data storage is probably corrupt."
+                ));
+            }
+            blocks.append(&mut read_blocks);
+        }
+        iroha_logger::info!("Read {} blocks from block store.", blocks.len());
+        Ok(blocks)
     }
+}
 
-    /// Returns a sorted vector of blocks starting from 0 height to the top block.
-    #[iroha_futures::telemetry_future]
-    pub async fn read_all(&self) -> Vec<VersionedCommittedBlock> {
-        let mut height = 1;
-        let mut blocks = Vec::new();
-        while let Ok(block) = self.read(height).await {
-            blocks.push(block);
-            height += 1;
+/// Returns deserialized blocks from specific file
+async fn read_all_from_storage_file(path: &Path) -> Result<Vec<VersionedCommittedBlock>> {
+    let mut blocks: Vec<VersionedCommittedBlock> = Vec::new();
+    let mut file = File::open(&path).await.wrap_err("No file found.")?;
+    let metadata = metadata(&path).await.wrap_err("Unable to read metadata.")?;
```

Comment:
Could you import `metadata` also as `fs::metadata`?

Reply:
> Could you import `metadata` also as `fs::metadata`?

std::fs or tokio::fs?

Reply:
`tokio::fs`